File: ports/ruby/lib/sprite_tool/sprite.rb
# frozen_string_literal: true

require "open3"
require "shellwords"

module SpriteTool
  # Wrapper around the sprite CLI, providing exec, file transfer,
  # checkpoint, and lifecycle methods.
  class Sprite
    attr_reader :name, :dry_run

    def initialize(name, dry_run: false)
      @name = name
      @dry_run = dry_run
    end

    # Run `sprite exec -s <sprite> bash -c "<cmd>"` and return stdout.
    # Raises on non-zero exit.
    def exec(cmd)
      if @dry_run
        puts "  [dry-run] sprite exec -s #{@name} bash -c #{cmd.shellescape}"
        return ""
      end
      out, status = Open3.capture2("sprite", "exec", "-s", @name, "bash", "-c", cmd)
      unless status.success?
        raise "sprite exec failed (exit #{status.exitstatus}): #{cmd}\n#{out}"
      end
      out.strip
    end

    # Run `sprite exec -s <sprite> bash -c "<cmd>"` piping stdin_data through stdin.
    def exec_with_stdin(cmd, stdin_data)
      if @dry_run
        puts "  [dry-run] sprite exec -s #{@name} bash -c #{cmd.shellescape} < (stdin)"
        return ""
      end
      out, status = Open3.capture2(
        "sprite", "exec", "-s", @name, "bash", "-c", cmd,
        stdin_data: stdin_data
      )
      unless status.success?
        raise "sprite exec with stdin failed (exit #{status.exitstatus}): #{cmd}\n#{out}"
      end
      out.strip
    end

    # Run `sprite exec -s <sprite> bash -c "<cmd>"` with stdin/stdout/stderr
    # connected to the terminal (interactive).
    def exec_interactive(cmd)
      if @dry_run
        puts "  [dry-run] sprite exec -s #{@name} bash -c #{cmd.shellescape}"
        return true
      end
      system("sprite", "exec", "-s", @name, "bash", "-c", cmd)
    end

    # Run `sprite exec -s <sprite> <args...>` with stdout piped to a writer,
    # stderr to terminal. Returns [stdout_read_io, wait_thread].
    def exec_piped_stdout(*args)
      full_args = ["sprite", "exec", "-s", @name] + args
      if @dry_run
        puts "  [dry-run] #{full_args.join(" ")}"
        return [StringIO.new(""), nil]
      end
      stdin_r, stdin_w = IO.pipe
      stdin_w.close
      stdout_r, stdout_w = IO.pipe
      pid = spawn(*full_args, in: stdin_r, out: stdout_w, err: $stderr)
      stdin_r.close
      stdout_w.close
      wait_thread = Process.detach(pid)
      [stdout_r, wait_thread]
    end

    # Run `sprite exec -s <sprite> bash -c "<cmd>"`, piping local stdin to remote
    # and remote stdout to local stdout. Used for tar streaming.
    def exec_pipe_through(cmd, stdin_io: nil, stdout_io: nil)
      if @dry_run
        puts "  [dry-run] sprite exec -s #{@name} bash -c #{cmd.shellescape}"
        return true
      end

      opts = {}
      opts[:in] = stdin_io if stdin_io
      opts[:out] = stdout_io if stdout_io
      opts[:err] = $stderr

      system("sprite", "exec", "-s", @name, "bash", "-c", cmd, **opts)
    end

    # Push a local file to the sprite at dest path.
    def push_file(src, dest)
      if @dry_run
        puts "  [dry-run] push #{src} -> sprite:#{dest}"
        return
      end
      dest_dir = File.dirname(dest)
      self.exec("mkdir -p #{dest_dir.shellescape}")
      data = File.binread(src)
      exec_with_stdin("cat > #{dest.shellescape}", data)
    end

    # Push a local directory to the sprite at dest path via tar.
    def push_dir(src, dest)
      if @dry_run
        puts "  [dry-run] push dir #{src} -> sprite:#{dest}"
        return
      end
      self.exec("mkdir -p #{dest.shellescape}")
      parent = File.dirname(src)
      base = File.basename(src)
      dest_parent = File.dirname(dest)

      # Use Open3.popen3 to pipe tar through sprite exec
      tar_cmd = ["tar", "czf", "-", "-C", parent, base]
      remote_cmd = "tar xzf - -C #{dest_parent.shellescape}"
      sprite_cmd = ["sprite", "exec", "-s", @name, "bash", "-c", remote_cmd]

      Open3.pipeline(tar_cmd, sprite_cmd)
    end

    # Run a plain `sprite <args...>` command and return stdout.
    def run(*args)
      if @dry_run
        puts "  [dry-run] sprite #{args.join(" ")}"
        return ""
      end
      out, status = Open3.capture2("sprite", *args)
      unless status.success?
        raise "sprite #{args.join(" ")} failed (exit #{status.exitstatus})\n#{out}"
      end
      out.strip
    end

    # Run a plain `sprite <args...>` command with terminal I/O.
    def run_passthrough(*args)
      if @dry_run
        puts "  [dry-run] sprite #{args.join(" ")}"
        return true
      end
      system("sprite", *args)
    end

    # Check if the sprite CLI is installed; install if not.
    def self.ensure_cli(dry_run: false)
      return if system("which sprite > /dev/null 2>&1")

      puts "sprite CLI not found, installing..."
      if dry_run
        puts "  [dry-run] curl -fsSL https://sprites.dev/install.sh | sh"
        return
      end
      system("bash", "-c", "curl -fsSL https://sprites.dev/install.sh | sh")
      ENV["PATH"] = "#{ENV["HOME"]}/.local/bin:#{ENV["PATH"]}"
    end

    # Authenticate with sprite using token or interactive login.
    def self.auth(token, dry_run: false)
      if token && !token.empty?
        puts "Authenticating sprite with token..."
        unless dry_run
          system("sprite", "auth", "setup", "--token", token)
        end
      else
        puts "No SPRITE_TOKEN set. Running interactive login..."
        unless dry_run
          system("sprite", "login")
        end
      end
    end

    # Check if this sprite already exists.
    def exists?
      if @dry_run
        puts "  [dry-run] sprite ls (checking for #{@name})"
        return false
      end
      out, status = Open3.capture2("sprite", "ls")
      return false unless status.success?
      out.split("\n").any? { |line| line.include?(@name) }
    end

    # Create the sprite with `sprite create -skip-console <name>`.
    def create
      if @dry_run
        puts "  [dry-run] sprite create -skip-console #{@name}"
        return
      end
      system("sprite", "create", "-skip-console", @name)
    end

    # Create a checkpoint of the sprite.
    def checkpoint_create
      if @dry_run
        puts "  [dry-run] sprite checkpoint create -s #{@name}"
        return true
      end
      system("sprite", "checkpoint", "create", "-s", @name)
    end

    # Open a console to the sprite.
    def console
      if @dry_run
        puts "  [dry-run] sprite console -s #{@name}"
        return
      end
      system("sprite", "console", "-s", @name)
    end
  end
end

File: ports/clojure/src/sprite_tool/sprite.clj
(ns sprite-tool.sprite
  "Subprocess wrapper for the sprite CLI using ProcessBuilder."
  (:require [clojure.string :as str])
  (:import [java.io File]
           [java.lang ProcessBuilder ProcessBuilder$Redirect]))

(defn find-sprite-cli
  "Check if the sprite CLI is on PATH. Returns true if found."
  []
  (try
    (let [pb (ProcessBuilder. ["which" "sprite"])]
      (.redirectErrorStream pb true)
      (let [proc (.start pb)
            exit (.waitFor proc)]
        (zero? exit)))
    (catch Exception _ false)))

(defn run-process
  "Run a command via ProcessBuilder. Returns the exit code.
   Options:
     :inherit-io  - inherit stdin/stdout/stderr (default true)
     :capture     - capture stdout as string (default false)
     :stdin-file  - redirect stdin from this file path
     :stdin-pipe  - a Process whose stdout should be piped to this stdin"
  [args & {:keys [inherit-io capture stdin-file stdin-pipe]
           :or {inherit-io true capture false}}]
  (let [pb (ProcessBuilder. ^java.util.List (vec args))]
    (when stdin-file
      (.redirectInput pb (File. ^String stdin-file)))
    (when inherit-io
      (when-not capture
        (if stdin-file
          ;; inheritIO would override stdin redirect, so only inherit stdout/stderr
          (doto pb
            (.redirectOutput ProcessBuilder$Redirect/INHERIT)
            (.redirectError ProcessBuilder$Redirect/INHERIT))
          (.inheritIO pb))))
    (when stdin-pipe
      ;; stdin-pipe is an InputStream from another process
      ;; We handle this after start
      nil)
    (let [proc (.start pb)]
      (when stdin-pipe
        ;; Copy from the input stream to process stdin
        (let [out (.getOutputStream proc)]
          (try
            (let [buf (byte-array 8192)]
              (loop []
                (let [n (.read ^java.io.InputStream stdin-pipe buf)]
                  (when (pos? n)
                    (.write out buf 0 n)
                    (recur)))))
            (finally
              (.close out)))))
      (if capture
        (let [stdout (slurp (.getInputStream proc))
              exit   (.waitFor proc)]
          {:exit exit :stdout (str/trim stdout)})
        (let [exit (.waitFor proc)]
          {:exit exit})))))

(defn sx
  "Run a command inside a sprite via bash.
   Equivalent to: sprite exec -s SPRITE bash -c CMD
   Returns captured stdout as string, or nil in dry-run mode."
  [sprite-name cmd & {:keys [dry-run] :or {dry-run false}}]
  (if dry-run
    (do
      (println (str "  [dry-run] sprite exec -s " sprite-name " bash -c \"" cmd "\""))
      nil)
    (let [result (run-process ["sprite" "exec" "-s" sprite-name "bash" "-c" cmd]
                              :inherit-io false :capture true)]
      (:stdout result))))

(defn sx-passthrough
  "Run a command inside a sprite, passing stdout/stderr through to terminal.
   Returns the exit code."
  [sprite-name cmd & {:keys [dry-run] :or {dry-run false}}]
  (if dry-run
    (do
      (println (str "  [dry-run] sprite exec -s " sprite-name " bash -c \"" cmd "\""))
      0)
    (let [result (run-process ["sprite" "exec" "-s" sprite-name "bash" "-c" cmd]
                              :inherit-io true)]
      (:exit result))))

(defn push-file
  "Push a local file to a sprite.
   Equivalent to:
     sprite exec -s SPRITE bash -c 'mkdir -p $(dirname DEST)'
     sprite exec -s SPRITE bash -c 'cat > DEST' < src"
  [sprite-name src dest & {:keys [dry-run] :or {dry-run false}}]
  (if dry-run
    (println (str "  [dry-run] push " src " -> sprite:" dest))
    (let [dest-dir (.getParent (File. ^String dest))
          dest-dir (or dest-dir "/")]
      ;; mkdir -p
      (run-process ["sprite" "exec" "-s" sprite-name "bash" "-c"
                    (str "mkdir -p '" dest-dir "'")]
                   :inherit-io true)
      ;; cat > dest < src
      (run-process ["sprite" "exec" "-s" sprite-name "bash" "-c"
                    (str "cat > '" dest "'")]
                   :inherit-io true :stdin-file src))))

(defn push-dir
  "Push a local directory to a sprite via tar.
   Equivalent to:
     sprite exec -s SPRITE bash -c 'mkdir -p DEST'
     tar czf - -C parent base | sprite exec -s SPRITE bash -c 'tar xzf - -C parent_of_dest'"
  [sprite-name src dest & {:keys [dry-run] :or {dry-run false}}]
  (if dry-run
    (println (str "  [dry-run] push dir " src " -> sprite:" dest))
    (let [src-file    (File. ^String src)
          parent      (.getParent src-file)
          base        (.getName src-file)
          dest-parent (or (.getParent (File. ^String dest)) "/")]
      ;; mkdir -p dest
      (run-process ["sprite" "exec" "-s" sprite-name "bash" "-c"
                    (str "mkdir -p '" dest "'")]
                   :inherit-io true)
      ;; tar czf - -C parent base | sprite exec ... bash -c "tar xzf - -C dest_parent"
      (let [tar-pb (doto (ProcessBuilder. ^java.util.List
                                          ["tar" "czf" "-" "-C" parent base])
                     (.redirectOutput ProcessBuilder$Redirect/PIPE)
                     (.redirectError ProcessBuilder$Redirect/INHERIT))
            tar-proc (.start tar-pb)

            sprite-pb (doto (ProcessBuilder. ^java.util.List
                                             ["sprite" "exec" "-s" sprite-name
                                              "bash" "-c"
                                              (str "tar xzf - -C '" dest-parent "'")])
                        (.redirectOutput ProcessBuilder$Redirect/INHERIT)
                        (.redirectError ProcessBuilder$Redirect/INHERIT))
            sprite-proc (.start sprite-pb)]
        ;; Pipe tar stdout -> sprite stdin in a thread
        (let [tar-out (.getInputStream tar-proc)
              sprite-in (.getOutputStream sprite-proc)
              pipe-thread (Thread.
                           (fn []
                             (try
                               (let [buf (byte-array 8192)]
                                 (loop []
                                   (let [n (.read tar-out buf)]
                                     (when (pos? n)
                                       (.write sprite-in buf 0 n)
                                       (recur)))))
                               (catch Exception _)
                               (finally
                                 (.close sprite-in)))))]
          (.start pipe-thread)
          (.join pipe-thread)
          (.waitFor tar-proc)
          (let [exit (.waitFor sprite-proc)]
            (when-not (zero? exit)
              (throw (ex-info "push-dir failed" {:exit exit})))))))))

(defn sprite-list
  "Run 'sprite ls' and return stdout."
  []
  (let [result (run-process ["sprite" "ls"] :inherit-io false :capture true)]
    (or (:stdout result) "")))

(defn sprite-exists?
  "Check if a sprite with the given name already exists."
  [sprite-name]
  (let [output (sprite-list)]
    (some (fn [line]
            (let [parts (str/split (str/trim line) #"\s+")]
              (some #(= % sprite-name) parts)))
          (str/split-lines output))))
